### PR TITLE
Update to create fully-featured Ruby SDK

### PR DESCRIPTION
**Pull request by jrweingart**
## Description of the change
1. Params for basic auth and setting base url
2. Env variables added for basic auth, token auth, and base url
2. Profile API, Events API, Lists API, Messages API, Brands API support:
Profiles:
- PATCH /profiles/{recipient_id}
- GET /profiles/{recipient_id}
- PUT /profiles/{recipient_id}
- POST /profiles/{recipient_id}
- GET /profiles/{recipient_id}/lists

Events:
- GET /events
- GET /events/{event_id}
- PUT /events/{event_id}

Lists:
- POST /send/list
- GET /profiles/{recipient_id}/lists
- GET /lists
- GET /lists/{list_id}
- PUT /lists/{list_id}
- DELETE /lists/{list_id}
- PUT /lists/{list_id}/restore
- GET /lists/{list_id}/subscriptions
- PUT /lists/{list_id}/subscriptions
- PUT /lists/{list_id}/subscriptions/{recipient_id}
- DELETE /lists/{list_id}/subscriptions/{recipient_id}

Messages:
- GET /messages
- GET /messages/{message_id}
- GET /messages/{message_id}/history

Brands:
- GET /brands
- GET /brands/{brand_id}
- POST /brands
- PUT /brands/brand_id
- DELETE /brands/brand_id

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review 

- [x]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: .github/gem-push.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      - name: Publish to RubyGems
+        run: |
+          mkdir -p $HOME/.gem
+          touch $HOME/.gem/credentials
+          chmod 0600 $HOME/.gem/credentials
+          printf -- "---\n:rubygems_api_key: ${GEM_HOST_API_KEY}\n" > $HOME/.gem/credentials
+          gem build *.gemspec
+          gem push *.gem
+        env:
+          GEM_HOST_API_KEY: "${{secrets.RUBYGEMS_AUTH_TOKEN}}"
```
Comment (aydrian) [inline .github/gem-push.yml:49]: Has this secret been added? I don't have access to it.

File: CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  - `DELETE /lists/{list_id}` with `client.lists.delete` method
+  - `PUT /lists/{list_id}/restore` with `client.lists.restore` method
+  - `GET /lists/{list_id}/subscriptions` with `client.lists.get_subscriptions` method
+  - `PUT /lists/{list_id}/subscriptions` with `client.lists.put_subscriptions` method
+  - `PUT /lists/{list_id}/subscriptions/{recipient_id}` with `client.lists.subscribe` method
+  - `DELETE /lists/{list_id}/subscriptions/{recipient_id}` with `client.lists.unsubscribe` method
+
+- Support for Profiles API by @jrweingart
+  - `GET /profiles/{recipient_id}` with `client.profiles.get` method
+  - `GET /profiles/{recipient_id}/lists` with `client.profiles.get_subscriptions` method
```
Comment (aydrian) [inline CHANGELOG.md:29]: This route is covered twice. Once in Profiles and another in List. Did we mean to do that?

File: lib/trycourier.rb [hunk 3]
```diff
@@ -18,49 +29,75 @@ def initialize(code, message_id)
...
+        @session.init_token_auth(auth_token)
+      elsif ENV["COURIER_AUTH_TOKEN"]
+        @session.init_token_auth(ENV["COURIER_AUTH_TOKEN"])
+      elsif username && password
+        @session.init_basic_auth(username, password)
+      elsif ENV["COURIER_AUTH_USERNAME"] && ENV["COURIER_AUTH_PASSWORD"]
+        @session.init_basic_auth(ENV["COURIER_AUTH_USERNAME"], ENV["COURIER_AUTH_PASSWORD"])
+      end
+
+      @messages = Courier::Messages.new(@base_url, @session)
```
Comment (aydrian) [inline lib/trycourier.rb:53]: Why is the `base_url` not part of the `Courier::CourierAPISession`? It feels like it should be included like the auth is and then we don't have to pass it around extra.

File: lib/trycourier.rb [hunk 4]
```diff
@@ -18,49 +29,75 @@ def initialize(code, message_id)
...
-      http = Net::HTTP.new(@uri.host, @uri.port)
-      http.use_ssl = true
-      http.verify_mode = OpenSSL::SSL::VERIFY_PEER
-
-      req = Net::HTTP::Post.new(@uri)
-      req["authorization"] = "Bearer #{@auth_token}"
-      req["content-type"] = "application/json"
-      req["User-Agent"] = "courier-ruby/#{Courier::VERSION}"
-      req.body = body.to_json
+      url = @base_url + "/send"
```
Comment (aydrian) [inline lib/trycourier.rb:73]: Here, if `base_url` was part of the session, then you would only need to pass `/send` and construct it in the session.

Comment (aydrian) [reply lib/trycourier.rb:73]: I do like how you cleaned this up and used a session.

File: lib/trycourier/brands.rb [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+      code = res.code.to_i
+      if code >= 400
+        obj = JSON.parse res.read_body
+        if (message = obj["Message"].nil? ? obj["message"] : obj["Message"])
+          err = "#{code}: #{message}"
+          raise CourierAPIException, err
+        end
+      end
+      res
+    end
```
Comment (aydrian) [inline lib/trycourier/brands.rb:40]: It feels like this logic shouldn't be duplicated in each endpoint class. Maybe we actually need to build out the `CourierAPIExpection` class instead of defining it as generic in each file.

File: lib/trycourier/brands.rb [hunk 6]
```diff
@@ -0,0 +1,105 @@
+require "trycourier/version"
+require "net/http"
+require "json"
+require "openssl"
+
+module Courier
+  # generic error to catch all exceptions
+  class CourierAPIException < StandardError; end
```
Comment (aydrian) [inline lib/trycourier/brands.rb:8]: I think we only need to do this once inside the Courier module. Feels like we are redefining it unnecessarily.

File: trycourier.gemspec [hunk 7]
```diff
@@ -1,15 +1,15 @@
...
 
 Gem::Specification.new do |spec|
-  spec.name          = "trycourier"
-  spec.version       = Courier::VERSION
-  spec.authors       = ["Courier"]
-  spec.email         = ["support@trycourier.com"]
+  spec.name = "trycourier"
+  spec.version = Courier::VERSION
+  spec.authors = ["Courier"]
+  spec.email = ["support@trycourier.com"]
```
Comment (aydrian) [inline trycourier.gemspec:7]: We need to make sure to get rid of all the trycourier urls (trycourier.app, trycourier.com)
```suggestion
spec.email = ["support@courier.com"]
```

**Review by aydrian — CHANGES_REQUESTED** [review]
A few things to look at.

File: CHANGELOG.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+### Added
+- Minor bug fixes to ensure proper SSL certification by @scarney81
+
+## v1.0.2 - 2021-1-6
+### Added
+- Support for Send API by @troy
+
+Initial release.
+
+[unreleased]: https://github.com/trycourier/courier-ruby-rails/compare/v0.1.3...HEAD
```
Comment (aydrian) [inline CHANGELOG.md:61]:
```suggestion
[unreleased]: https://github.com/trycourier/courier-ruby/compare/v1.0.2...HEAD
[v1.0.2]: https://github.com/trycourier/courier-ruby/compare/v1.0.1...v1.0.2
[v1.0.1]: https://github.com/trycourier/courier-ruby/compare/v1.0.0...v1.0.1
```
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  - `PUT /events/{event_id}` with `client.events.replace` and `client.events.add` methods
+
+- Support for Brands API by @jrweingart
+  - `GET /brands` with `client.brands.list` method
+  - `GET /brands/brand_id` with `client.brands.get` method
+  - `POST /brands` with `client.brands.create` method
+  - `PUT /brands/brand_id` with `client.brands.replace` method
+  - `DELETE /brands/brand_id` with `client.brands.delete` method
+
+## v1.0.1 - 2020-3-3
```
Comment (aydrian) [inline CHANGELOG.md:51]:
```suggestion
## [v1.0.2] - 2021-1-6
```

File: CHANGELOG.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+### Added
+- Minor bug fixes to ensure proper SSL certification by @scarney81
+
+## v1.0.2 - 2021-1-6
```
Comment (aydrian) [inline CHANGELOG.md:55]:
```suggestion
## [v1.0.1] - 2020-3-4
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+### Added
+- Support for Send API by @troy
+
+Initial release.
```
Comment (aydrian) [inline CHANGELOG.md:59]:
```suggestion
## v1.0.0 - 2020-3-3
Initial release by @troygoode 
```

**Review by aydrian — CHANGES_REQUESTED** [review]
Change log fixes.

File: spec/lists_spec.rb [hunk 12]
```diff
@@ -33,7 +33,7 @@
           headers: TOKEN_AUTH_HEADERS
         )
         .to_return(body: "{\"status\": \"ok\"}", status: 200)
-      res = client.lists.send(NOTIFICATION_ID, list: LIST_ID, brand: BRAND_ID, override: {"provider": {}})
+      res = client.lists.send(NOTIFICATION_ID, list_id: LIST_ID, brand: BRAND_ID, override: {"provider": {}})
```
Comment (aydrian) [inline spec/lists_spec.rb:36]: Just noticed this. We should only use `list_id` if it matches the API. Here it should be list `list`. There are some places in the Lists API where it is `list_id`. It just gets confusing. We have some issues with it in the node library.

**Review by aydrian — CHANGES_REQUESTED** [review]

File: lib/trycourier/brands.rb [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        "settings" => settings
+      }
+      if id
+        payload["id"] = id
+      end
+      if snippets
+        payload["snippets"] = snippets
+      end
+
+      res = @session.send(@@key, "POST", body: payload, headers: headers)
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/brands.rb:45]:
```suggestion
      res = @session.send(KEY, "POST", body: payload, headers: headers)
```

File: lib/trycourier/brands.rb [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+      ErrorHandler.check_err(res)
+    end
+
+    def replace(brand_id:, name:, settings:, snippets: nil)
+      path = @@key + "/" + brand_id
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/brands.rb:50]:
```suggestion
      path = "#{KEY}/#{brand_id}"
```

File: lib/trycourier/brands.rb [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+      if snippets
+        payload["snippets"] = snippets
+      end
+
+      res = @session.send(path, "PUT", body: payload)
+      ErrorHandler.check_err_non_json(res)
+    end
+
+    def delete(brand_id:)
+      path = @@key + "/" + brand_id
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/brands.rb:65]:
```suggestion
      path = "#{KEY}/#{brand_id}"
```

File: lib/trycourier/events.rb [hunk 16]
```diff
@@ -0,0 +1,35 @@
+module Courier
+  class Events
+    @@key = "/events"
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/events.rb:3]:
```suggestion
    KEY = "/events"
```

File: lib/trycourier/events.rb [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+    def initialize(session)
+      @session = session
+    end
+
+    def list
+      res = @session.send(@@key, "GET")
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/events.rb:10]:
```suggestion
      res = @session.send(KEY, "GET")
```

File: lib/trycourier/events.rb [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+      ErrorHandler.check_err(res)
+    end
+
+    def get(event_id:)
+      path = @@key + "/" + event_id.to_s
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/events.rb:15]:
```suggestion
      path = "#{KEY}/#{event_id}"
```

File: lib/trycourier/events.rb [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+      res = @session.send(path, "GET")
+      ErrorHandler.check_err(res)
+    end
+
+    def add(event_id:, id:, type: "notification")
+      replace(event_id: event_id, id: id, type: type)
+    end
+
+    def replace(event_id:, id:, type: "notification")
+      path = @@key + "/" + event_id.to_s
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/events.rb:25]:
```suggestion
      path = "#{KEY}/#{event_id}"
```

File: lib/trycourier/exceptions.rb [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+  class InputError < StandardError; end
+  class ResponseError < StandardError; end
+  class CourierAPIError < ResponseError; end  
+
+  class ErrorHandler
+    def self.check_err(res)
+      code = res.code.to_i
+      obj = JSON.parse res.read_body
+
+      if code < 400
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/exceptions.rb:17]: This would run this code receiving a 3xx response. I highly doubt an API would ever redirect you, I mean it could, but what sane one would ever? lol. I would check for explicit status codes here, but its not a blocker.

File: lib/trycourier/lists.rb [hunk 21]
```diff
@@ -0,0 +1,109 @@
+module Courier
+  class Lists
+    @@key = "/lists"
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/lists.rb:3]:
```suggestion
    KEY = "/lists"
```

File: lib/trycourier/lists.rb [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+
+    def list(cursor: nil, pattern: nil)
+      params = {}
+      if cursor
+        params["cursor"] = cursor
+      end
+      if pattern
+        params["pattern"] = pattern
+      end
+      res = @session.send(@@key, "GET", params: params)
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/lists.rb:45]:
```suggestion
      res = @session.send(KEY, "GET", params: params)
```

File: lib/trycourier/lists.rb [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+      ErrorHandler.check_err(res)
+    end
+
+    def get(list_id:)
+      path = @@key + "/" + list_id.to_s
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/lists.rb:50]:
```suggestion
      path = "#{KEY}/#{list.id}"
```

File: lib/trycourier/lists.rb [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+      res = @session.send(path, "GET")
+      ErrorHandler.check_err(res)
+    end
+
+    def put(list_id:, name:)
+      path = @@key + "/" + list_id.to_s
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/lists.rb:56]:
```suggestion
      path = "#{KEY}/#{list.id}"
```

File: lib/trycourier/lists.rb [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+      payload = {
+        "name": name.to_s
+      }
+
+      res = @session.send(path, "PUT", body: payload)
+      ErrorHandler.check_err_non_json(res)
+    end
+
+    def delete(list_id:)
+      path = @@key + "/" + list_id.to_s
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/lists.rb:67]:
```suggestion
      path = "#{KEY}/#{list.id}"
```

File: lib/trycourier/lists.rb [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+      res = @session.send(path, "DELETE")
+      ErrorHandler.check_err_non_json(res)
+    end
+
+    def restore(list_id:)
+      path = @@key + "/" + list_id.to_s + "/restore"
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/lists.rb:73]:
```suggestion
      path = "#{KEY}/#{list.id}/restore"
```

File: lib/trycourier/lists.rb [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+      res = @session.send(path, "PUT")
+      ErrorHandler.check_err_non_json(res)
+    end
+
+    def get_subscriptions(list_id:, cursor: nil)
+      path = @@key + "/" + list_id.to_s + "/subscriptions"
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/lists.rb:79]:
```suggestion
      path = "#{KEY}/#{list.id}/subscriptions"
```

File: lib/trycourier/lists.rb [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+      params = {}
+      if cursor
+        params["cursor"] = cursor
+      end
+      res = @session.send(path, "GET", params: params)
+      ErrorHandler.check_err(res)
+    end
+
+    def put_subscriptions(list_id:, recipients:)
+      path = @@key + "/" + list_id.to_s + "/subscriptions"
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/lists.rb:89]:
```suggestion
      path = "#{KEY}/#{list.id}/subscriptions"
```

File: lib/trycourier/lists.rb [hunk 29]
```diff
@@ -0,0 +1,109 @@
...
+      payload = {
+        "recipients": recipients
+      }
+      res = @session.send(path, "PUT", body: payload)
+      ErrorHandler.check_err_non_json(res)
+    end
+
+    def subscribe(list_id:, recipient_id:)
+      path = @@key + "/" + list_id.to_s + "/subscriptions/" + recipient_id.to_s
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/lists.rb:98]:
```suggestion
      path = "#{KEY}/#{list.id}/subscriptions/recipient_id"
```

File: lib/trycourier/lists.rb [hunk 30]
```diff
@@ -0,0 +1,109 @@
...
+      res = @session.send(path, "PUT")
+      ErrorHandler.check_err_non_json(res)
+    end
+
+    def unsubscribe(list_id:, recipient_id:)
+      path = @@key + "/" + list_id.to_s + "/subscriptions/" + recipient_id.to_s
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/lists.rb:104]:
```suggestion
      path = "#{KEY}/#{list.id}/subscriptions/recipient_id"
```

File: lib/trycourier/messages.rb [hunk 31]
```diff
@@ -0,0 +1,51 @@
+module Courier
+  class Messages
+    @@key = "/messages"
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/messages.rb:3]:
```suggestion
    KEY = "/messages"
```

File: lib/trycourier/messages.rb [hunk 32]
```diff
@@ -0,0 +1,51 @@
...
+      if message_id
+        params["messageId"] = message_id
+      end
+      if notification
+        params["notification"] = notification
+      end
+      if recipient
+        params["recipient"] = recipient
+      end
+      res = @session.send(@@key, "GET", params: params)
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/messages.rb:31]:
```suggestion
      res = @session.send(KEY, "GET", params: params)
```

File: lib/trycourier/messages.rb [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+      ErrorHandler.check_err(res)
+    end
+
+    def get(message_id:)
+      path = @@key + "/" + message_id.to_s
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/messages.rb:36]:
```suggestion
      path = "#{KEY}/#{message_id}"
```

File: lib/trycourier/messages.rb [hunk 34]
```diff
@@ -0,0 +1,51 @@
...
+      res = @session.send(path, "GET")
+      ErrorHandler.check_err(res)
+    end
+
+    def get_history(message_id:, type: nil)
+      path = @@key + "/" + message_id.to_s + "/history"
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/messages.rb:42]:
```suggestion
      path = "#{KEY}/#{message_id}/history"
```

File: lib/trycourier/profiles.rb [hunk 35]
```diff
@@ -0,0 +1,64 @@
+module Courier
+  class Profiles
+    @@key = "/profiles"
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/profiles.rb:3]:
```suggestion
    KEY = "/profiles"
```

File: lib/trycourier/profiles.rb [hunk 36]
```diff
@@ -0,0 +1,64 @@
...
+
+    def initialize(session)
+      @session = session
+    end
+
+    def get(recipient_id:)
+      path = @@key + "/" + recipient_id.to_s
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/profiles.rb:10]:
```suggestion
      path = "#{KEY}/#{recipient_id}"
```

File: lib/trycourier/profiles.rb [hunk 37]
```diff
@@ -0,0 +1,64 @@
...
+      res = @session.send(path, "GET")
+      ErrorHandler.check_err(res)
+    end
+
+    def get_subscriptions(recipient_id:, cursor: nil)
+      path = @@key + "/" + recipient_id.to_s + "/subscriptions"
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/profiles.rb:16]:
```suggestion
      path = "#{KEY}/#{recipient_id}/subscriptions"
```

File: lib/trycourier/profiles.rb [hunk 38]
```diff
@@ -0,0 +1,64 @@
...
+      res = @session.send(path, "GET", params: params)
+      ErrorHandler.check_err(res)
+    end
+
+    def add(recipient_id:, profile:)
+      replace(recipient_id: recipient_id, profile: profile)
+    end
+
+    def replace(recipient_id:, profile:)
+      path = @@key + "/" + recipient_id.to_s
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/profiles.rb:32]:
```suggestion
      path = "#{KEY}/#{recipient_id}"
```

File: lib/trycourier/profiles.rb [hunk 39]
```diff
@@ -0,0 +1,64 @@
...
+      payload = {
+        'profile': profile
+      }
+
+      res = @session.send(path, "PUT", body: payload)
+      ErrorHandler.check_err_non_json(res)
+    end
+
+    def merge(recipient_id:, profile:, idempotency_key: nil)
+      path = @@key + "/" + recipient_id.to_s
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/profiles.rb:43]:
```suggestion
      path = "#{KEY}/#{recipient_id}"
```

File: lib/trycourier/profiles.rb [hunk 40]
```diff
@@ -0,0 +1,64 @@
...
+      headers = {}
+      if idempotency_key
+        headers["Idempotency-Key"] = idempotency_key
+      end
+      res = @session.send(path, "POST", body: payload, headers: headers)
+      ErrorHandler.check_err(res)
+    end
+
+    def patch(recipient_id:, operations:)
+      path = @@key + "/" + recipient_id.to_s
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/profiles.rb:56]:
```suggestion
      path = "#{KEY}/#{recipient_id}"
```

File: lib/trycourier/session.rb [hunk 41]
```diff
@@ -0,0 +1,85 @@
...
+
+module Courier
+  class CourierAPISession
+
+    def initialize(base_url)
+      @base_url = base_url
+      @auth_token = nil
+      @username = nil
+      @password = nil
+      @auth_method = nil
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/session.rb:13]: An instance variable in ruby will by default evaluate to nil, so these are unnecessary.

```suggestion
```

File: lib/trycourier/session.rb [hunk 42]
```diff
@@ -0,0 +1,85 @@
...
+      if headers
+        headers.each do |k, v|
+          req.add_field(k.to_s, v.to_s)
+        end
+      end
+
+      http.request(req)
+    end
+
+    def isAuthenticated
```
Comment (notactuallytreyanastasio) [inline lib/trycourier/session.rb:74]: I would recommend making this method `is_authenticated` as its a public method and while you might not expect a user to call this, they probably someday might want to, and in Ruby method names are snake cased conventionally. It would be a lot of suggestions to make it an automagic change, but I would recommend either making this method private, putting a TODO to change its name, and ensure your users can't use it, or rename it and keep it public.

File: spec/events_spec.rb [hunk 43]
```diff
@@ -0,0 +1,110 @@
+require_relative "spec_helper"
+
+RSpec.describe Courier::Events do
+  let(:client) { Courier::Client.new(AUTH_TOKEN_MOCK) }
```
Comment (notactuallytreyanastasio) [inline spec/events_spec.rb:4]: I would refrain from putting constants that are autoloaded into the global namespace by requiring spec_helper, but it works well enough for your needs here. To work around it elegantly you'd need some sort of factory solution and that is out of scope for this work.

File: spec/messages_spec.rb [hunk 44]
```diff
@@ -0,0 +1,83 @@
...
+                                 notification: "78", message_id: "91", list_id: "23")
+      expect(res).to eq({"paging" => {}, "items" => []})
+    end
+
+    it "fails with exception" do
+      stub_request(:get, "https://api.courier.com/messages")
+        .with(
+          headers: TOKEN_AUTH_HEADERS
+        ).to_return(body: "{\"message\": \"an error occurred\"}", status: 400)
+      expect { client.messages.list }.to raise_error(Courier::CourierAPIError)
```
Comment (notactuallytreyanastasio) [inline spec/messages_spec.rb:33]: Nice-to-have, but you can also chain `.with_message` to `raise_error` in all of these and ensure that your error messaging is accurate on top of ensuring it raises the expected error. Far from essential but its a nice feature I leverage a lot myself with custom exceptions for library code.

**Review by notactuallytreyanastasio — COMMENTED** [review]
It looks like a ton but I mainly tried to make it so I could allow all the suggestions for moving away from class variables batchable so you wouldn't have to go through and manually change them all if you wanted to accept the request. Overall this looks good! There are a few things that are not quite 'conventional ruby' but considering the needs of this work it would be out of scope to request all that kind of nitpicking.

One thing I would mention, I notice that there is a move from using symbol based hashes to exclusively string based hashes here. You could get around this by including `active_support` and using the `with_indifferent_access` method, which allows you to access the keys via string _or_ symbol. However, `active_support` is a *big* dependency, and I wouldn't change to that style unless you were in a position where a user got some unexpected behavior by using symbol based hashes. The docs seem pretty clear though so who knows if that would happen.

File: spec/messages_spec.rb (hunk 44 above)
Comment (jrweingart) [reply spec/messages_spec.rb:33]: Thank you for all of the feedback!

File: .github/gem-push.yml (hunk 1 above)
Comment (DannyDouglass) [reply .github/gem-push.yml:49]: @aydrian it has not been added. send it my way and i'll add.

Comment (DannyDouglass) [reply .github/gem-push.yml:49]: added:
image

File: lib/trycourier/brands.rb [hunk 45]
```diff
@@ -0,0 +1,70 @@
...
+      ErrorHandler.check_err(res)
+    end
+
+    def get(brand_id:)
+      path = @@key + "/" + brand_id.to_s
+      res = @session.send(path, "GET")
+      ErrorHandler.check_err(res)
+    end
+
+    def add(name:, settings:, id: nil, snippets: nil, idempotency_key: nil)
```
[the remark this answers is not in the document]
Comment (jrweingart) [reply lib/trycourier/brands.rb:24]: Doesn't this syntax make name and settings required fields?

**Review by DannyDouglass — APPROVED** [review]

**Review by aydrian — APPROVED** [review]

**Comment by DannyDouglass** [comment]
@notactuallytreyanastasio thank you for all the feedback - very much appreciated! the other comments you left we are going to capture in another ticket to round back to in the near future.

please let us know if we can return the favor  ❤️

File: lib/trycourier/brands.rb (hunk 45 above)
[the remark this answers is not in the document]
Comment (notactuallytreyanastasio) [reply lib/trycourier/brands.rb:24]: It does -- which I would consider a feature and not a bug. There is a high amount of required KWargs throughout this that strengthen the contracts between all the method signatures, where if they didn't default to nil they would lead to you chasing down nil. The benefit is that you get an explicit, clear error over a nil bubbling through. If you really dont want them to be required, I would remove the presence of kwargs as a whole here because defaulting them to nil functionally gets you to the same point of having to chase it down later over see the clear exception.
